### PR TITLE
[VLM] feat: true on policy for vlm + fsdp

### PR DESCRIPTION
**Pull request by nanjiangwill**
## Motivation

Support true on policy for VLM + FSDP

Exp: Qwen3-VL

Text-only: [#12058](https://github.com/sgl-project/sglang/pull/12058) 

## Modifications

when `get_global_server_args().rl_on_policy_target is not None`
1. change `MRotaryEmbedding.forward()` to be `forward_native()` 
2. PIL -> `pixel_values` conversion under `cpu` not `cuda`
3. `position_embeddings` to be same type as `inputs` in ViT
6. change `fast_pos_embed_interpolate` from `numpy` to `torch` implementation
7. set env `SGLANG_VLM_CACHE_SIZE_MB=0` 
8. add `input_deepstack_embeds` to `hidden_states` after `hidden_states + residual` since `hidden_states + residual + input_deepstack_embeds != hidden_states + input_deepstack_embeds + residual`

## Checklist

- [X] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

File: python/sglang/srt/layers/rotary_embedding.py [hunk 1]
```diff
@@ -1510,14 +1511,12 @@ def forward(
...
-        elif _is_npu:
-            return self._forward_npu(positions, query, key)
-        else:
-            return self._forward_native(positions, query, key)
+        # Use Triton kernel for multimodal (2D positions) with mrope
+        if positions.ndim == 2 and self.mrope_section:
+            return self.forward_triton(positions, query, key)
+        return self.forward_native(positions, query, key, fused_set_kv_buffer_arg)
 
-    def _forward_triton(
```
Comment (zhaochenyang20) [inline python/sglang/srt/layers/rotary_embedding.py:1553]: no need to change it from `_forward` to `forward`?

File: python/sglang/srt/layers/rotary_embedding.py [hunk 2]
```diff
@@ -1566,15 +1565,19 @@ def _forward_triton(
         key = torch.cat((key_rot, key_pass), dim=-1).reshape(key_shape)
         return query, key
 
-    def _forward_npu(
```
Comment (zhaochenyang20) [inline python/sglang/srt/layers/rotary_embedding.py:1574]: no need to change it from _forward to forward?

File: python/sglang/srt/layers/rotary_embedding.py (hunk 1 above)
Comment (nanjiangwill) [reply python/sglang/srt/layers/rotary_embedding.py:1553]: this inherits RotaryEmbedding and has CustomOp, we should just keep on standard. so we just follow the format in class RotaryEmbedding

File: python/sglang/srt/layers/rotary_embedding.py (hunk 2 above)
Comment (nanjiangwill) [reply python/sglang/srt/layers/rotary_embedding.py:1574]: same rationale


this inherits RotaryEmbedding and has CustomOp, we should just keep on standard. so we just follow the format in class RotaryEmbedding

File: python/sglang/srt/layers/layernorm.py [hunk 3]
```diff
@@ -121,6 +121,10 @@ def forward_cuda(
                 self.variance_epsilon,
             )
         if residual is not None:
+            # TODO: Ideally we want to have (a+b)+c. but right now we can only have a+(b+c).
+            # (a+b)+c != a+(b+c), we probably need to add another parameter to fused_add_rmsnorm
```
Comment (zhaochenyang20) [inline python/sglang/srt/layers/layernorm.py]: what is a b c here stands for?

**Review by zhaochenyang20 — CHANGES_REQUESTED** [review]

File: python/sglang/srt/layers/layernorm.py (hunk 3 above)
Comment (nanjiangwill) [reply python/sglang/srt/layers/layernorm.py]: updated!

**Review by zhaochenyang20 — APPROVED** [review]
LGTM.

**Comment by mickqian** [comment]
could you attach the mmmu accuracy change in here? since this PR modifies core implementation of qwen3-vl

**Review by mickqian — APPROVED** [review]

File: python/sglang/srt/layers/layernorm.py [hunk 4]
```diff
@@ -179,13 +190,23 @@ def forward_native(
...
-            x = x + residual.to(torch.float32)
+            x = (
+                x
+                + residual.to(torch.float32)
+                + (
+                    post_residual_addition.to(torch.float32)
+                    if post_residual_addition is not None
+                    else 0.0
+                )
+            )
```
Comment (JustinTong0323) [inline python/sglang/srt/layers/layernorm.py:210]: Maybe it is cleaner to write in this way:
```suggestion
        post_residual_addition = kwargs.get("post_residual_addition", 0.0).to(torch.float32)
        x = x.to(torch.float32)
        if residual is not None:
            x = x + residual.to(torch.float32)
            x = (
                x
                + residual.to(torch.float32)
                + post_residual_addition
            )
```

Comment (nanjiangwill) [reply python/sglang/srt/layers/layernorm.py:210]: we need to make to(float) inside the if. but i will make the default value 0.0

File: python/sglang/srt/multimodal/processors/base_processor.py [hunk 5]
```diff
@@ -318,7 +319,9 @@ def process_mm_data(
             and isinstance(processor.image_processor, BaseImageProcessorFast)
             and not self.server_args.disable_fast_image_processor
         ):
-            if not _is_npu:
+            if get_global_server_args().rl_on_policy_target is not None:
+                kwargs["device"] = "cpu"
```
Comment (JustinTong0323) [inline python/sglang/srt/multimodal/processors/base_processor.py:321]: Could you please explain more about this? Why using fast processor would break onpolicy🤔

**Comment by JustinTong0323** [comment]
Also please resolve the conflicts~

File: python/sglang/srt/multimodal/processors/base_processor.py (hunk 5 above)
Comment (nanjiangwill) [reply python/sglang/srt/multimodal/processors/base_processor.py:321]: because of the image processing, it will generate `pixel_values` in float, and doing that on GPU will give mismatch between Sglang and HF

**Comment by yxchng** [comment]
is this going to be merged in new version soon?

File: python/sglang/srt/server_args.py [hunk 6]
```diff
@@ -2329,6 +2329,9 @@ def _handle_deterministic_inference(self):
                 "Enable deterministic inference because of rl_on_policy_target."
             )
             self.enable_deterministic_inference = True
+
+            # For VLM
+            os.environ["SGLANG_VLM_CACHE_SIZE_MB"] = "0"
```
Comment (JustinTong0323) [inline python/sglang/srt/server_args.py:2329]: why default it to 0?

Comment (nanjiangwill) [reply python/sglang/srt/server_args.py:2329]: this will only be triggered when using true on policy mode. we need to evict image cache for now as it will introduce mismatch. will optimize this in the future.

Comment (JustinTong0323) [reply python/sglang/srt/server_args.py:2329]: Thanks for the explanation, that makes sense.

**Review by JustinTong0323 — APPROVED** [review]

File: python/sglang/srt/layers/layernorm.py [hunk 7]
```diff
@@ -105,21 +105,30 @@ def forward_cuda(
...
                 self.variance_epsilon,
             )
         if residual is not None:
+            # TODO: Ideally we want to have (hidden_states+residual)+post_residual_addition.
+            # but right now we can only have hidden_states+(residual+post_residual_addition).
+            # (hidden_states+residual)+post_residual_addition != hidden_states+(residual+post_residual_addition),
+            # we probably need to add another parameter to fused_add_rmsnorm
+            post_residual_addition = kwargs.get("post_residual_addition")
+            residual = residual + (
+                post_residual_addition if post_residual_addition is not None else 0.0
```
Comment (merrymercy) [inline python/sglang/srt/layers/layernorm.py:130]: This adds unnecessary overhead. Please be super careful next time. https://github.com/sgl-project/sglang/pull/16474/changes

Also, please fix all other places where you did the same mistake

File: python/sglang/srt/layers/layernorm.py [hunk 8]
```diff
@@ -105,21 +105,30 @@ def forward_cuda(
         self,
         x: torch.Tensor,
         residual: Optional[torch.Tensor] = None,
+        **kwargs,
```
Comment (merrymercy) [inline python/sglang/srt/layers/layernorm.py:108]: `kwargs` is a very bad pattern. Use it unless you do not have any other methods to deal with it.
Why is this needed?

File: python/sglang/srt/layers/layernorm.py [hunk 9]
```diff
@@ -179,13 +191,23 @@ def forward_native(
...
         x = x.to(torch.float32)
         if residual is not None:
-            x = x + residual.to(torch.float32)
+            x = (
+                x
+                + residual.to(torch.float32)
+                + (
+                    post_residual_addition.to(torch.float32)
+                    if post_residual_addition is not None
+                    else 0.0
```
Comment (merrymercy) [inline python/sglang/srt/layers/layernorm.py:208]: Do not add unnecessary 0.0
